app/models/miq_set_membership.rb

```diff
@@ -0,0 +1,4 @@
+class MiqSetMembership < ApplicationRecord
```

Comment:
I think membership needs to be sti, and unfortunately we need to introduce 10 subclasses.
It is no secret that I am very much opposed to sti.

We can introduce multiple tables. (please no)
We can have polymorphic links to set and members. (this is rails friendly but not sql friendly).

We could be clever and use where and cast calls from each of the callers. but we would also need to introduce miq_set getters/setters here.

having polymorphic with each child class defining the belongs_to with the proper subclass makes the most sense to me since it is straight forward, very rails friendly, and very sql friendly. (It just unfortunately uses STI and is repetitive).

We could auto generate the child classes but that introduces bugs with the autoloader.

Reply:
MiqSet uses STI (for the sets themselves)
MiqSetMembership is already polymorphic - https://github.com/ManageIQ/manageiq-schema/pull/536/files#diff-507b248cbcc874533629f4be6fa2a7b8763df2657c2c27efd806a07540d7978cR4

This is what I was referring to here - https://github.com/ManageIQ/manageiq/pull/20665#discussion_r501397343. We could eliminate miq_sets table and have 20 different miq_set shaped tables, but that feels terrible.

Reply:
sets all point to the same table. so the set_id is unique.
The set type class is ambiguous, but when we pull back the record it will figure out the proper class via a plugin/STI

when we are going from set to member, there is sti to get the right object back, but this will force us to read the membership records, figure out the member class, and then fetch the records.

if the set pointed to a membership class that specifically stated the membership record class, then rails could just do a join and you would not need to use polymorphism and `includes` would be available to you.

when fetching all MiqSet records with an includes, that would get tricky. but when fetching all custom buttons sets and members, then includes would be able to use proper sql constructs.

Reply:
Not sure how much my view on this one has changed. polymorphism may generate the same queries as STI.